File: Frontend/banka-app/src/app/utils/internal-transfer-detection.ts
import { Transaction } from '../models/transaction.model';

/** Detecta IDs de transferencias internas entre cuentas propias (mismo día, mismo importe, signos opuestos) */
export function detectInternalTransferIds(txs: { dt_date?: string; importe?: number; cuenta?: string; transaction_id?: string; id?: number }[]): Set<string> {
  const INTERNAL_ACCOUNTS = new Set(txs.map(t => (t.cuenta || '').trim()).filter(Boolean));
  const matched = new Set<string>();
  const byKey = new Map<string, typeof txs>();
  for (const t of txs) {
    const cuenta = (t.cuenta || '').trim();
    if (!INTERNAL_ACCOUNTS.has(cuenta)) continue;
    const imp = t.importe ?? 0;
    if (imp === 0) continue;
    const fecha = t.dt_date || '';
    if (!fecha) continue;
    const key = `${fecha}|${Math.abs(Math.round(imp * 100) / 100)}`;
    if (!byKey.has(key)) byKey.set(key, []);
    byKey.get(key)!.push(t);
  }
  for (const [, group] of byKey) {
    const negatives = group.filter(t => (t.importe ?? 0) < 0);
    const positives = group.filter(t => (t.importe ?? 0) > 0);
    if (!negatives.length || !positives.length) continue;
    const usedPos = new Set<number>();
    for (const neg of negatives) {
      const cNeg = (neg.cuenta || '').trim();
      for (let i = 0; i < positives.length; i++) {
        if (usedPos.has(i)) continue;
        if ((positives[i].cuenta || '').trim() !== cNeg) {
          matched.add(neg.transaction_id || String(neg.id || ''));
          matched.add(positives[i].transaction_id || String(positives[i].id || ''));
          usedPos.add(i);
          break;
        }
      }
    }
  }
  return matched;
}

File: Backend/app/api/services/pipe_extract_transactions/internal_transfer_detection.py
"""
Detección de transferencias internas entre cuentas propias.
Cuando se detecta: mismo día, mismo importe absoluto, signo opuesto, cuentas distintas
(Revolut, Personal, Conjunta), se marca como transferencia interna para que:
- No afecte a gastos/ingresos/balances
- Se muestre en un apartado dedicado "Transferencias internas"
"""
from typing import Set
from collections import defaultdict

INTERNAL_ACCOUNTS = frozenset({"Revolut", "Personal", "Conjunta"})


def detect_internal_transfer_ids(transactions: list[dict]) -> Set[str]:
    """
    Detecta pares de transferencias internas y retorna los IDs/transaction_ids
    de todas las transacciones que forman parte de un par.

    Criterios: mismo día, mismo importe absoluto, una negativa y otra positiva,
    cuentas distintas dentro de {Revolut, Personal, Conjunta}.
    """
    if not transactions:
        return set()

    def get_id(t: dict) -> str:
        return t.get("transaction_id") or t.get("id") or str(t.get("id", ""))

    def get_date(t: dict) -> str:
        return str(t.get("dt_date") or t.get("transaction_date") or "")

    def get_importe(t: dict) -> float:
        v = t.get("importe") if "importe" in t else t.get("amount")
        try:
            return float(v) if v is not None else 0.0
        except (TypeError, ValueError):
            return 0.0

    def get_cuenta(t: dict) -> str:
        return str(t.get("cuenta") or t.get("account_number") or "").strip()

    matched_ids: Set[str] = set()

    # Agrupar por (fecha, importe_absoluto)
    by_key: dict[tuple[str, float], list[dict]] = defaultdict(list)
    for t in transactions:
        cuenta = get_cuenta(t)
        if cuenta not in INTERNAL_ACCOUNTS:
            continue
        imp = get_importe(t)
        if imp == 0:
            continue
        fecha = get_date(t)
        if not fecha:
            continue
        key = (fecha, abs(round(imp, 2)))
        by_key[key].append(t)

    for (fecha, amount_abs), group in by_key.items():
        negatives = [t for t in group if get_importe(t) < 0]
        positives = [t for t in group if get_importe(t) > 0]
        if not negatives or not positives:
            continue
        # Emparejar: cada negativa con una positiva de cuenta distinta
        used_pos = set()
        for neg in negatives:
            c_neg = get_cuenta(neg)
            for i, pos in enumerate(positives):
                if i in used_pos:
                    continue
                if get_cuenta(pos) != c_neg:
                    matched_ids.add(get_id(neg))
                    matched_ids.add(get_id(pos))
                    used_pos.add(i)
                    break

    return matched_ids
